src/Adapter/MSTest.TestAdapter/Execution/TestClassInfo.cs

```diff
@@ -343,18 +345,29 @@ public void RunClassInitialize(TestContext testContext)
 
     internal UnitTestResult GetResultOrRunClassInitialize(ITestContext testContext, string initializationLogs, string initializationErrorLogs, string initializationTrace, string initializationTestContextMessages)
     {
+        // Optimization: If we already ran before and know the result, return it.
+        if (_classInitializeResult is not null)
+        {
+            DebugEx.Assert(IsClassInitializeExecuted, "Class initialize result should be available if and only if class initialize was executed");
+            return _classInitializeResult;
+        }
+
+        DebugEx.Assert(!IsClassInitializeExecuted, "If class initialize was executed, we should have been in the previous if were we have a result available.");
+
         bool isSTATestClass = AttributeComparer.IsDerived<STATestClassAttribute>(ClassAttribute);
         bool isWindowsOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
         if (isSTATestClass
             && isWindowsOS
             && Thread.CurrentThread.GetApartmentState() != ApartmentState.STA)
         {
-            // For optimization purposes, we duplicate some of the logic of RunClassInitialize here so we don't need to start
-            // a thread for nothing.
-            if ((ClassInitializeMethod is null && BaseClassInitMethods.Count == 0)
-                || IsClassInitializeExecuted)
+            // For optimization purposes, avoid starting a thread when we know it will do nothing.
+            // This is the case when:
+            // 1. We have already executed class initialize (this is handled at the beginning for STA and non-STA)
+            // 2. There is no class initialize method to execute. In this case we mark IsClassInitializeExecuted and return a passed result (this condition here).
+            if (ClassInitializeMethod is null && BaseClassInitMethods.Count == 0)
```

Comment:
Looking from the phone so I have a less good view but can't we also move this condition early in the method? It seems independent to using sta thread or not.

Reply:
Indeed. That makes sense.